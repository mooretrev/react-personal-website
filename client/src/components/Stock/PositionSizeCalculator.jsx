import React, { useState } from 'react';
import TextField from '@material-ui/core/TextField';
import Card from '@material-ui/core/Card';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import PropTypes from 'prop-types';
import useStyles from '../../styles/Index.jsx';

function PositionSizeCalculator(props) {
  const classes = useStyles();

  const [accountSize, setAccountSize] = useState('');
  const [stopLoss, setStopLoss] = useState('');
  const [entry, setEntry] = useState('');

  const handleAccountSizeInput = (event) => {
    setAccountSize(event.target.value);
  };
  const handleStopLossInput = (event) => {
    setStopLoss(event.target.value);
  };
  const handleEntryInput = (event) => {
    setEntry(event.target.value);
  };

  const displayPositionSize = () => {
    if (accountSize !== '' && stopLoss !== '' && entry !== '') {
      const accountSizeInt = parseInt(accountSize, 10);
      const stopLossInt = parseInt(stopLoss, 10);
      const entryInt = parseInt(entry, 10);

      const amountRisk = (accountSizeInt * props.risk);
      const numShares = amountRisk / (entryInt - stopLossInt);
      const size = entryInt * numShares;
      return (
        <Typography id="output" variant="h6">
          Position Size is
          {' '}
          {size}
          .
          {' '}
          <br />
          Number of Shares is
          {' '}
          {numShares}
          .
          {' '}
          <br />
          Amount at risk
          {' '}
          {amountRisk}
          .
        </Typography>
      );
    }
    return <div />;
  };

  return (
    <div>
      <Card>
        <Grid container className={classes.cardPadding}>
          <Grid item className={classes.gridItemSpacing} xs={12}>
            <Typography variant="h2">
              Position Size Calculator
            </Typography>
          </Grid>
          <Grid item className={classes.gridItemSpacing} xs={12} md={4}>
            <TextField fullWidth value={accountSize} id="acct_size_input" label="Account Size" variant="outlined" onChange={handleAccountSizeInput} />
          </Grid>
          <Grid item className={classes.gridItemSpacing} xs={12} md={4}>
            <TextField fullWidth value={stopLoss} id="stop_loss_input" label="Stop Loss" variant="outlined" onChange={handleStopLossInput} />
          </Grid>
          <Grid item className={classes.gridItemSpacing} xs={12} md={4}>
            <TextField fullWidth value={entry} id="entry_input" label="Entry" variant="outlined" onChange={handleEntryInput} />
          </Grid>
          <Grid item className={classes.gridItemSpacing} xs={12}>
            {displayPositionSize()}
          </Grid>
        </Grid>
      </Card>
    </div>
  );
}

PositionSizeCalculator.propTypes = {
  risk: PropTypes.number.isRequired,
};

export default PositionSizeCalculator;
