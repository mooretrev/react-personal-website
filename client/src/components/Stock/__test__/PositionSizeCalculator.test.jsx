import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import PositionSizeCalculator from '../PositionSizeCalculator.jsx';

describe('<PositionSizeCalculator />', () => {
  it('should render correctly', () => {
    const recipeForm = (
      <PositionSizeCalculator risk={0.02} />
    );
    const wrapper = shallow(recipeForm);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should take in values and purpose the correct output', () => {
    const recipeForm = (
      <PositionSizeCalculator risk={0.02} />
    );
    const wrapper = shallow(recipeForm);

    let acctSizeInput = wrapper.find('#acct_size_input');
    acctSizeInput.simulate('change', { target: { value: '3000' } });
    acctSizeInput = wrapper.find('#acct_size_input');
    expect(acctSizeInput.props().value).toBe('3000');

    let stopLossInput = wrapper.find('#stop_loss_input');
    stopLossInput.simulate('change', { target: { value: '50' } });
    stopLossInput = wrapper.find('#stop_loss_input');
    expect(stopLossInput.props().value).toBe('50');

    let entryInput = wrapper.find('#entry_input');
    entryInput.simulate('change', { target: { value: '55' } });
    entryInput = wrapper.find('#entry_input');
    expect(entryInput.props().value).toBe('55');

    const output = wrapper.find('#output');
    expect(output.html().includes('Position Size is 660.')).toBe(true);
    expect(output.html().includes('Number of Shares is 12.')).toBe(true);
    expect(output.html().includes('Amount at risk 60.')).toBe(true);
  });
});
