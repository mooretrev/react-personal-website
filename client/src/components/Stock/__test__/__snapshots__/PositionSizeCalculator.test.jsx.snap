// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PositionSizeCalculator /> should render correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Card))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-cardPadding-13"
      container={true}
    >
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-gridItemSpacing-14"
        item={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          variant="h2"
        >
          Position Size Calculator
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-gridItemSpacing-14"
        item={true}
        md={4}
        xs={12}
      >
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          id="acct_size_input"
          label="Account Size"
          onChange={[Function]}
          value=""
          variant="outlined"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-gridItemSpacing-14"
        item={true}
        md={4}
        xs={12}
      >
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          id="stop_loss_input"
          label="Stop Loss"
          onChange={[Function]}
          value=""
          variant="outlined"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-gridItemSpacing-14"
        item={true}
        md={4}
        xs={12}
      >
        <WithStyles(ForwardRef(TextField))
          fullWidth={true}
          id="entry_input"
          label="Entry"
          onChange={[Function]}
          value=""
          variant="outlined"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-gridItemSpacing-14"
        item={true}
        xs={12}
      >
        <div />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Card))>
</div>
`;
